### PR TITLE
Batching

### PR DESCRIPTION
**Pull request by jdoliner**

### CODE REVIEW

**Comment by gabrielgrant** [comment]
Is this implementing batching from the user's POV, or just in the sending of datums to workers (ie with no API change)?

**Comment by jdoliner** [comment]
So far this PR has no API changes. I think I'm likely going to merge this in and then do any API changes in a separate PR.

File: src/server/pkg/pool/pool.go [hunk 1]
```diff
@@ -70,65 +68,72 @@ func (p *Pool) watchEndpoints() {
...
+				if err != nil {
+					return err
+				}
+				conn = &connCount{cc: cc}
+				p.conns[addr] = conn
+				// We break because this conn has a count of 0 which we know
+				// we're not beating
+				break
+			} else {
+				if conn == nil || mapConn.count < conn.count {
```
Comment (derekchiang) [inline src/server/pkg/pool/pool.go]: I think for atomic variables, you are supposed to use an `atomic.Load*` function to get the value of it.

File: src/server/pkg/pool/pool.go [hunk 2]
```diff
@@ -70,65 +68,72 @@ func (p *Pool) watchEndpoints() {
...
 	}
+	defer atomic.AddInt64(&conn.count, -1)
+	return f(conn.cc)
 }
 
 // Close closes all connections stored in the pool, it returns an error if any
 // of the calls to Close error.
 func (p *Pool) Close() error {
 	close(p.done)
-	close(p.conns)
```
Comment (derekchiang) [inline src/server/pkg/pool/pool.go]: hype

File: src/server/pkg/worker/master.go [hunk 3]
```diff
@@ -112,7 +113,7 @@ func (a *APIServer) jobSpawner(ctx context.Context, logger *taggedLogger) error
 	if err != nil {
 		return err
 	}
-	pool, err := pool.NewPool(a.kubeClient, a.namespace, ppsserver.PipelineRcName(a.pipelineInfo.Pipeline.Name, a.pipelineInfo.Version), numWorkers, client.PachDialOptions()...)
+	pool, err := pool.NewPool(a.kubeClient, a.namespace, ppsserver.PipelineRcName(a.pipelineInfo.Pipeline.Name, a.pipelineInfo.Version), numWorkers*5, client.PachDialOptions()...)
```
Comment (derekchiang) [inline src/server/pkg/worker/master.go]: Can we make this magic number a constant?

File: src/server/pkg/worker/master.go [hunk 4]
```diff
@@ -388,7 +390,7 @@ func (a *APIServer) runJob(ctx context.Context, jobInfo *pps.JobInfo, pool *pool
 		}
 
 		failed := false
-		limiter := limit.New(a.numWorkers)
+		limiter := limit.New(a.numWorkers * 10)
```
Comment (derekchiang) [inline src/server/pkg/worker/master.go]: Not sure if I'm missing something, but why is this 10 while the other number is 5?

File: src/client/pps/pps.proto [hunk 5]
```diff
@@ -178,6 +179,7 @@ message JobInfo {
   pfs.Commit stats_commit = 29;
   bool enable_stats = 32;
   string salt = 33;
+  bool batch = 34;
```
Comment (jdoliner) [inline src/client/pps/pps.proto]: I doesn't actually implement the logic (as you've no doubt seen) I added this flag before I split off the other branch to implement user space batching and figured it wasn't really worth rewriting git history to get rid of it.

File: src/server/pkg/pool/pool.go [hunk 6]
```diff
@@ -70,65 +68,72 @@ func (p *Pool) watchEndpoints() {
...
 		// According the k8s docs, the full set of endpoints is the cross
 		// product of (addresses x ports).
 		for _, address := range subset.Addresses {
 			for _, port := range subset.Ports {
-				addresses = append(addresses, fmt.Sprintf("%s:%d", address.IP, port.Port))
+				addr := fmt.Sprintf("%s:%d", address.IP, port.Port)
+				if cc := p.conns[addr]; cc != nil {
+					addresses[addr] = cc
+				} else {
+					addresses[addr] = nil
```
Comment (jdoliner) [inline src/server/pkg/pool/pool.go]: I like your idea so I'll just do that, and add a few comments as well.

File: src/server/pkg/worker/master.go (hunk 3 above)
Comment (jdoliner) [inline src/server/pkg/worker/master.go]: Actually we can do something even better which is make this number go away all together.

File: src/server/pkg/worker/master.go (hunk 4 above)
Comment (jdoliner) [inline src/server/pkg/worker/master.go]: This one we'll make into a constant though.

File: src/server/pkg/pool/pool.go [hunk 7]
```diff
@@ -12,19 +13,18 @@ import (
 	"k8s.io/kubernetes/pkg/watch"
 )
 
-// Pool stores a pool of grpc connections to, it's useful in places where you
-// would otherwise need to create several connections.
+type connCount struct {
```
Comment (jdoliner) [inline src/server/pkg/pool/pool.go]: Comments added.

**Comment by jdoliner** [comment]
@derekchiang all comments should be addressed, another look whenever you get a chance.

**Pull request merged**